### ISSUE TITLE
Problem with snap.exe of Chessmaster 3000 Dos

### ISSUE THREAD

**Issue by pinop76** [joncampbell123/dosbox-x#2646]
Hi,

there is a compatibility problem with dosbox-X and snap.exe (utility of game Chessmaster 3000 Dos version for system info id est cpu ram, ems, xms etc.), When snap.exe is running the system is blocked.
By classic dosbox the utility is ok. Why?

**Comment by Wengier** [comment]
I cannot seem to find the said executable in my downloaded copy of Chessmaster 3000. Where did you download it?

**Comment by pinop76** [comment]
> I cannot seem to find the said executable in my downloaded copy of Chessmaster 3000. Where did you download it?

I did not download it. I have an original copy of Chessmania that has Chessmaster 3000 Dos 1.04 version, Sargon 4 and Grandmaster Chess Dos. Do you want send the files of snap.exe?

**Comment by Wengier** [comment]
@pinop76 Sure, you can upload the file here. Thanks.

**Comment by pinop76** [comment]
[SNAP.zip](https://github.com/joncampbell123/dosbox-x/files/6787360/SNAP.zip)
Hi,
excuse me for delay but I was busy.  In classic dosbox snap.exe run very well, in dosbox-x no.
Verify, please. In my opinion there is a bug in dosbox-x about video function.

**Comment by grapeli** [comment]
It works. You need to disable XMS.
`dosbox-x -set "dos xms=false" -c "c:" -c "snap" SNAP.zip`

**Comment by pinop76** [comment]
> It works. You need to disable XMS.
> `dosbox-x -set "dos xms=false" -c "c:" -c "snap" SNAP.zip`

It don't function. And in classic dosbox snap.exe run also with xms (snap.exe say also the xms present in system).

**Comment by Wengier** [comment]
@pinop76 Try set `hma=false` then. Both the game and snap should run with it.

**Comment by pinop76** [comment]
> @pinop76 Try set `hma=false` then. Both the game and snap should run with it.

Thanks. snap.exe run, but it must run also hma=true. It seems a bug of dosbox.
Snap.exe says also if hma is on or off!

**Comment by Wengier** [comment]
@pinop76 The option "hma" is an DOSBox-Xaddition; there is no such option "hma" in vanilla DOSBox, which is always fixed to "false". The default memory settings in DOSBox-X and vanilla DOSBox are apparently different.

**Comment by pinop76** [comment]
@Wengier, ok hma option is surplus in dosbox-X but snap.exe must run also when hma is enabled as in real computer.
In my opinion it's a bug because snap.exe recognize if hma is abilited or disabilited in system.

**Comment by joncampbell123** [comment]
"HMA" is meant to emulate that real DOS machines with XMS reserve 64KB - 16 bytes at the 1MB boundary. If DOSBox-X is still not emulating calls related to HMA properly, then that needs to be fixed.

**Comment by Wengier** [comment]
FYI, you can also get SNAP to run with the setting `dos=low,umb` (instead of `dos=high,umb`), or with the command `A20GATE OFF`. Alternatively, commenting out line 832-838 of `src/dos/dos.cpp` will also make it work. Seems that something is wrong with the XMS_DOS_LocalA20EnableIfNotEnabled() function.

**Comment by cimarronm** [comment]
Based upon your last comment Wengier, I'm pretty sure I know exactly what the problem is. The problem has nothing to do with dosbox's implementation but rather a bug in the packer used for the executable. If it is using Microsoft's EXEPACK or some others like PKLite, it has a wraparound bug in its depacking algorithm which relies on the code beginning execution above 64KiB. By moving dos out of high memory, the executable is staring above 64 KiB or by disabling A20 the wrap-around will work.

**Comment by joncampbell123** [comment]
@cimarronm Right, if that is the case, running LOADFIX prior to SNAP.EXE should allow it to run without crashing.

**Comment by grapeli** [comment]
Under FreeDOS 1.2 in Qemu, the snap will behave the same.
There is additional information - Invalid Opcode at ...
![snap freedos](https://user-images.githubusercontent.com/452325/125119998-5dbf1780-e0e1-11eb-8ca0-9f42b10e5cde.png)

**Comment by pinop76** [comment]
> Based upon your last comment Wengier, I'm pretty sure I know exactly what the problem is. The problem has nothing to do with dosbox's implementation but rather a bug in the packer used for the executable. If it is using Microsoft's EXEPACK or some others like PKLite, it has a wraparound bug in its depacking algorithm which relies on the code beginning execution above 64KiB. By moving dos out of high memory, the executable is staring above 64 KiB or by disabling A20 the wrap-around will work.

Then Its not abug of Dosbox but a bug of packer of snap.exe?
